File: reference/matlab/test_var.m
% TEST_VAR  Script-based test suite for the MATLAB VaR implementation.
%
% Run with:
%   cd reference/matlab
%   run('test_var.m')
%
% Prints "All tests passed." on success. Calls error() on failure.

this_dir     = fileparts(mfilename('fullpath'));
fixtures_dir = fullfile(this_dir, '..', 'fixtures');

returns = csvread(fullfile(fixtures_dir, 'sample_returns.csv'), 1, 0);
fprintf('Loaded %d returns\n\n', length(returns));

% ---------------------------------------------------------------------------
% Historical VaR
% ---------------------------------------------------------------------------

fprintf('Test 1: historical VaR is positive ... ');
[v, ~] = var_historical(returns);
assert(v > 0);
fprintf('PASS\n');

fprintf('Test 2: historical VaR monotonic in confidence ... ');
[v90, ~] = var_historical(returns, 0.90);
[v95, ~] = var_historical(returns, 0.95);
[v99, ~] = var_historical(returns, 0.99);
assert(v90 < v95 && v95 < v99);
fprintf('PASS\n');

fprintf('Test 3: historical CVaR >= VaR ... ');
for conf = [0.90, 0.95, 0.99]
    [v, cv] = var_historical(returns, conf);
    assert(cv >= v, 'CVaR < VaR at conf=%g', conf);
end
fprintf('PASS\n');

% ---------------------------------------------------------------------------
% Parametric VaR
% ---------------------------------------------------------------------------

fprintf('Test 4: parametric VaR is positive ... ');
[v, ~] = var_parametric(returns);
assert(v > 0);
fprintf('PASS\n');

fprintf('Test 5: parametric VaR monotonic in confidence ... ');
[v90, ~] = var_parametric(returns, 0.90);
[v95, ~] = var_parametric(returns, 0.95);
[v99, ~] = var_parametric(returns, 0.99);
assert(v90 < v95 && v95 < v99);
fprintf('PASS\n');

fprintf('Test 6: parametric CVaR >= VaR ... ');
for conf = [0.90, 0.95, 0.99]
    [v, cv] = var_parametric(returns, conf);
    assert(cv >= v, 'CVaR < VaR at conf=%g', conf);
end
fprintf('PASS\n');

% ---------------------------------------------------------------------------
% Monte Carlo VaR
% ---------------------------------------------------------------------------

fprintf('Test 7: MC VaR is positive ... ');
mc = var_monte_carlo(returns);
assert(mc.var > 0);
fprintf('PASS\n');

fprintf('Test 8: MC CVaR >= VaR ... ');
assert(mc.cvar >= mc.var);
fprintf('PASS\n');

fprintf('Test 9: MC seed reproducibility ... ');
mc1 = var_monte_carlo(returns, 0.95, 1, 10000, 123);
mc2 = var_monte_carlo(returns, 0.95, 1, 10000, 123);
assert(mc1.var == mc2.var);
assert(mc1.cvar == mc2.cvar);
fprintf('PASS\n');

% ---------------------------------------------------------------------------
% Fixture validation - historical
% ---------------------------------------------------------------------------

fprintf('Test 10: historical matches fixture values ... ');
expected = readtable(fullfile(fixtures_dir, 'expected_historical.csv'));
for i = 1:height(expected)
    [v, cv] = var_historical(returns, expected.confidence(i), expected.holding_period(i));
    assert(abs(v - expected.var(i)) / expected.var(i) < 0.01, ...
        'hist VaR mismatch at row %d', i);
    assert(abs(cv - expected.cvar(i)) / expected.cvar(i) < 0.01, ...
        'hist CVaR mismatch at row %d', i);
end
fprintf('PASS\n');

% ---------------------------------------------------------------------------
% Fixture validation - parametric
% ---------------------------------------------------------------------------

fprintf('Test 11: parametric matches fixture values ... ');
expected = readtable(fullfile(fixtures_dir, 'expected_parametric.csv'));
for i = 1:height(expected)
    [v, cv] = var_parametric(returns, expected.confidence(i), expected.holding_period(i));
    assert(abs(v - expected.var(i)) / expected.var(i) < 0.001, ...
        'param VaR mismatch at row %d', i);
    assert(abs(cv - expected.cvar(i)) / expected.cvar(i) < 0.001, ...
        'param CVaR mismatch at row %d', i);
end
fprintf('PASS\n');

fprintf('\n11/11 tests passed\n');

File: reference/r/test_var.R
#!/usr/bin/env Rscript
# Tests for the R VaR implementation.
# Run with: Rscript test_var.R

library(testthat)

# Source var.R relative to this script's location.
this_dir <- tryCatch(
  dirname(normalizePath(sys.frame(0)$ofile)),
  error = function(e) {
    args <- commandArgs(trailingOnly = FALSE)
    file_arg <- grep("^--file=", args, value = TRUE)
    if (length(file_arg) > 0) {
      dirname(normalizePath(sub("^--file=", "", file_arg)))
    } else {
      getwd()
    }
  }
)

source(file.path(this_dir, "var.R"))

FIXTURES_DIR <- normalizePath(file.path(this_dir, "..", "fixtures"))

# ---------------------------------------------------------------------------
# Fixture loading
# ---------------------------------------------------------------------------

load_returns <- function() {
  df <- read.csv(file.path(FIXTURES_DIR, "sample_returns.csv"))
  df$return
}

load_expected <- function(filename) {
  read.csv(file.path(FIXTURES_DIR, filename))
}

# ---------------------------------------------------------------------------
# log_returns
# ---------------------------------------------------------------------------

test_that("log_returns has correct length", {
  prices <- c(100, 105, 103, 107)
  lr <- log_returns(prices)
  expect_equal(length(lr), length(prices) - 1)
})

test_that("log_returns has correct values", {
  prices <- c(100, 110)
  lr <- log_returns(prices)
  expect_equal(lr[1], log(1.1))
})

# ---------------------------------------------------------------------------
# Historical Simulation
# ---------------------------------------------------------------------------

test_that("historical VaR is positive", {
  r <- load_returns()
  expect_gt(var_historical(r), 0)
})

test_that("historical VaR increases with confidence", {
  r <- load_returns()
  v90 <- var_historical(r, 0.90)
  v95 <- var_historical(r, 0.95)
  v99 <- var_historical(r, 0.99)
  expect_lt(v90, v95)
  expect_lt(v95, v99)
})

test_that("historical VaR increases with holding period", {
  r <- load_returns()
  v1  <- var_historical(r, holding_period = 1)
  v5  <- var_historical(r, holding_period = 5)
  v10 <- var_historical(r, holding_period = 10)
  expect_lt(v1, v5)
  expect_lt(v5, v10)
})

test_that("historical CVaR >= VaR", {
  r <- load_returns()
  for (conf in c(0.90, 0.95, 0.99)) {
    v  <- var_historical(r, conf)
    cv <- cvar_historical(r, conf)
    expect_gte(cv, v)
  }
})

# ---------------------------------------------------------------------------
# Parametric
# ---------------------------------------------------------------------------

test_that("parametric VaR is positive", {
  r <- load_returns()
  expect_gt(var_parametric(r), 0)
})

test_that("parametric VaR increases with confidence", {
  r <- load_returns()
  v90 <- var_parametric(r, 0.90)
  v95 <- var_parametric(r, 0.95)
  v99 <- var_parametric(r, 0.99)
  expect_lt(v90, v95)
  expect_lt(v95, v99)
})

test_that("parametric CVaR >= VaR", {
  r <- load_returns()
  for (conf in c(0.90, 0.95, 0.99)) {
    v  <- var_parametric(r, conf)
    cv <- cvar_parametric(r, conf)
    expect_gte(cv, v)
  }
})

# ---------------------------------------------------------------------------
# Monte Carlo
# ---------------------------------------------------------------------------

test_that("MC VaR is positive", {
  r <- load_returns()
  mc <- var_monte_carlo(r)
  expect_gt(mc$var, 0)
})

test_that("MC CVaR >= VaR", {
  r <- load_returns()
  mc <- var_monte_carlo(r, 0.95)
  expect_gte(mc$cvar, mc$var)
})

test_that("MC is seed-reproducible", {
  r <- load_returns()
  mc1 <- var_monte_carlo(r, seed = 123)
  mc2 <- var_monte_carlo(r, seed = 123)
  expect_equal(mc1$var, mc2$var)
  expect_equal(mc1$cvar, mc2$cvar)
})

test_that("MC returns correct number of simulations", {
  r <- load_returns()
  mc <- var_monte_carlo(r, n_simulations = 5000)
  expect_equal(length(mc$simulated_returns), 5000)
})

# ---------------------------------------------------------------------------
# Fixture validation - historical
# ---------------------------------------------------------------------------

test_that("historical matches fixture values", {
  r <- load_returns()
  expected <- load_expected("expected_historical.csv")
  for (i in seq_len(nrow(expected))) {
    row <- expected[i, ]
    v  <- var_historical(r, row$confidence, row$holding_period)
    cv <- cvar_historical(r, row$confidence, row$holding_period)
    expect_equal(v, row$var, tolerance = 0.01,
                 label = sprintf("hist VaR conf=%.2f hp=%d", row$confidence, row$holding_period))
    expect_equal(cv, row$cvar, tolerance = 0.01,
                 label = sprintf("hist CVaR conf=%.2f hp=%d", row$confidence, row$holding_period))
  }
})

# ---------------------------------------------------------------------------
# Fixture validation - parametric
# ---------------------------------------------------------------------------

test_that("parametric matches fixture values", {
  r <- load_returns()
  expected <- load_expected("expected_parametric.csv")
  for (i in seq_len(nrow(expected))) {
    row <- expected[i, ]
    v  <- var_parametric(r, row$confidence, row$holding_period)
    cv <- cvar_parametric(r, row$confidence, row$holding_period)
    expect_equal(v, row$var, tolerance = 0.001,
                 label = sprintf("param VaR conf=%.2f hp=%d", row$confidence, row$holding_period))
    expect_equal(cv, row$cvar, tolerance = 0.001,
                 label = sprintf("param CVaR conf=%.2f hp=%d", row$confidence, row$holding_period))
  }
})
